doc/viper/viper.md

```diff
@@ -0,0 +1,298 @@
+1. How to `Get` it out of a viper.
+
+`Configure`, therefore, has the following signature:
+
+```go
+func Configure[T any](key string, options Options[T]) Value[T]
+```
+
+The first parameter provides the key name (point 1 of our above list); all other information is provided via various `Options` fields, which looks like:
+
+```go
+type Options[T any] struct {
+    // what "things" to bind to
+    Aliases []string
+    FlagName string
+    EnvVars []string
+
+    // default, if any
+    Default T
+
+    // whether it can reload or not (more on this later)
+    Dynamic bool
+
+    // how to "get" it from a viper (more on this slightly less later)
+    GetFunc func(v *viper.Viper) func(key string) T
+}
+```
+
+### `Get` funcs
+
+In most cases, module authors will not need to specify a `GetFunc` option, since, if not provided, `viperutil` will do its best to provide a sensible default for the given type `T`.
+
+This requires a fair amount of `reflect`ion code, which we won't go into here, and unfortunately cannot support even all primitive types (notably, array (not slice!!) types).
+In these cases, the `GetFuncForType` will panic, allowing the module author to catch this during testing of their package.
+They may then provide their own `GetFunc`.
+
+Authors may also want to provide their own `GetFunc` to provide additional logic to load a value even for types supported by `GetFuncForType` (for example, post-processing a string to ensure it is always lowercase).
+
+The full suite of types, both supported and panic-inducing, are documented by way of unit tests in [`go/viperutil/get_func_test.go`](../../go/viperutil/get_func_test.go).
+
+### Dynamic values
+
+Values can be configured to be either static or dynamic.
+Static values are loaded once at startup (more precisely, when `viperutil.LoadConfig` is called), and whatever value is loaded at the point will be the result of calling `Get` on that value for the remainder of the processes lifetime.
+Dynamic values, conversely, may respond to config changes.
+
+In order for dynamic configs to be truly dynamic, `LoadConfig` must have found a config file (as opposed to pulling values entirely from defaults, flags, and environment variables).
+If this is the case, a second viper shim, which backs the dynamic registry, will start a watch on that file, and any changes to that file will be reflected in the `Get` methods of any values configured with `Dynamic: true`.
+
+**An important caveat** is that viper on its own is not threadsafe, meaning that if a config reload is being processed while a value is being accessed, a race condition can occur.
+To protect against this, the dynamic registry uses a second shim, [`sync.Viper`](../../go/viperutil/internal/sync/sync.go).
+This works by assigning each dynamic value its own `sync.RWMutex`, and locking it for writes whenever a config change is detected. Value `GetFunc`s are then adapted to wrap the underlying get in a `m.RLock(); defer m.RUnlock()` layer.
```

Comment:
RWMutex should only be used if the rate of reads is much higher than the rate of writes. Do we have a sense for whether that is expected to be the case?

Reply:
my intuition says yes because:
- reads happen anytime we are, well, reading the value. if we're constantly reading a given value, this number is going to be high. if we read sparingly (gated with some "check if the value changed" interval), then the number of reads is probably low enough that the RWMutex / Mutex tradeoff is probably negligible
- writes happen _only_ when the user is writing out a modified config file, so it's probably extremely low